### PR TITLE
add some basic async unit tests for the inflator/deflator streams

### PR DESCRIPTION
**Pull request by Numpsy**
refs #223 

Basically just adds async versions of the 'InflateDeflateZlib' and 'InflateDeflateNonZlib' tests and shuffles some things about to share some of the test code instead of duplicating it all.

Also changes the tests to use the NUnit 'Range' attribute for testing the different compression levels instead of having one test that loops over all the levels internally (makes the test a bit simpler, also makes each level appear as a seperate test in the results)

_I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License._

### CODE REVIEW

**Comment by Numpsy** [comment]
> 
> Also changes the tests to use the NUnit 'Range' attribute for testing the different compression levels instead of having one test that loops over all the levels internally (makes the test a bit simpler, also makes each level appear as a seperate test in the results)

Could break that bit out into a separate PR to minimize the diff on the actual async bits if that's helpful.

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 1]
```diff
@@ -89,35 +118,73 @@ private MemoryStream Deflate(byte[] data, int level, bool zlib)
...
+
+		/// <summary>
+		/// Basic async inflate/deflate test
+		/// </summary>
+		[Test]
+		[Category("Base")]
+		[Category("Async")]
+		public async Task InflateDeflateZlibAsync([Range(0, 9)] int level)
+		{
+			await RandomDeflateInflateAsync(100000, level, true);
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:179]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

**Comment by piksel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 6
           

Complexity decreasing per file
==============================
+ test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs  -3
         

Clones added
============
- test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476&bid=18799035)

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 2]
```diff
@@ -167,12 +234,20 @@ private void TryManyVariants(int level, bool zlib, RunCompress test, byte[] buff
...
+
+		/// <summary>
+		/// Basic async inflate/deflate test
+		/// </summary>
+		[Test]
+		[Category("Base")]
+		[Category("Async")]
+		public async Task InflateDeflateNonZlibAsync([Range(0, 9)] int level)
+		{
+			await RandomDeflateInflateAsync (100000, level, false);
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:250]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 3]
```diff
@@ -89,35 +118,73 @@ private MemoryStream Deflate(byte[] data, int level, bool zlib)
 			return memoryStream;
 		}
 
-		private void RandomDeflateInflate(int size, int level, bool zlib)
+		private async Task<MemoryStream> DeflateAsync(byte[] data, int level, bool zlib)
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:121]: Codacy found an issue: [Make 'DeflateAsync' a static method.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 4]
```diff
@@ -89,35 +118,73 @@ private MemoryStream Deflate(byte[] data, int level, bool zlib)
...
+
 			MemoryStream ms = Deflate(buffer, level, zlib);
 			Inflate(ms, buffer, level, zlib);
 		}
 
+		private async Task RandomDeflateInflateAsync(int size, int level, bool zlib)
+		{
+			byte[] buffer = GetRandomTestData(size);
+
+			MemoryStream ms = await DeflateAsync(buffer, level, zlib);
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:157]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 5]
```diff
@@ -15,19 +16,50 @@ namespace ICSharpCode.SharpZipLib.Tests.Base
 	[TestFixture]
 	public class InflaterDeflaterTestSuite
 	{
-		private void Inflate(MemoryStream ms, byte[] original, int level, bool zlib)
+		private void VerifyInflatedData(byte[] original, byte[] buf2, int level, bool zlib)
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:19]: Codacy found an issue: [Make 'VerifyInflatedData' a static method.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

File: test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs [hunk 6]
```diff
@@ -89,35 +118,73 @@ private MemoryStream Deflate(byte[] data, int level, bool zlib)
...
 			MemoryStream ms = Deflate(buffer, level, zlib);
 			Inflate(ms, buffer, level, zlib);
 		}
 
+		private async Task RandomDeflateInflateAsync(int size, int level, bool zlib)
+		{
+			byte[] buffer = GetRandomTestData(size);
+
+			MemoryStream ms = await DeflateAsync(buffer, level, zlib);
+			await InflateAsync(ms, buffer, level, zlib);
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Base/InflaterDeflaterTests.cs:158]: Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/piksel/SharpZipLib/pullRequest?prid=5745476)

**Comment by Numpsy** [comment]
I put the simple not-actually-async-related changes in #494 in case that makes it easier to review (could just close that one if not)

**Review by piksel — APPROVED** [review]
This was really hard to CR since it moved code in several places (and added new which was similar). I Tried to move around it a bit to get a better diff, but it's still a mess.
AFAICT it looks good though, it's just adding async versions of test helpers and moving shared sections to separate helpers.

**Comment by Numpsy** [comment]
Yes, there is perhaps a lot of scope for tidying some of these things up. (we might hit similar situations when/if adding async variants of the other existing tests later)

**Pull request merged**